From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
Date: Mon, 12 Oct 2026 07:54:52 +0200
Subject: [PATCH] Fit in the memory of the DS

The game's edicts and the client's entities come from the far heap, the
Memory Expansion Pak of a DS, the game's through the new TagMallocFar of
the game import, which Z_TagMallocFar takes from first. That RAM ignores
byte writes, they only have words. The directory of a pak is read there
instead of into 256 KiB of stack. 1024 particles instead of 4096.
---
 client/cl_main.c  |  6 ++++--
 client/client.h   |  2 +-
 client/ref.h      |  2 +-
 game/g_save.c     |  4 ++--
 game/game.h       |  3 +++
 qcommon/common.c  | 19 +++++++++++++++++--
 qcommon/files.c   |  4 +++-
 qcommon/qcommon.h |  1 +
 server/sv_game.c  |  1 +
 9 files changed, 33 insertions(+), 9 deletions(-)

diff --git a/client/cl_main.c b/client/cl_main.c
index e19423a..7bc0968 100644
--- a/client/cl_main.c
+++ b/client/cl_main.c
@@ -90,7 +90,8 @@ cvar_t	*cl_vwep;
 client_static_t	cls;
 client_state_t	cl;
 
-centity_t		cl_entities[MAX_EDICTS];
+// 276 KiB of words, in the far heap
+centity_t		*cl_entities;
 
 entity_state_t	cl_parse_entities[MAX_PARSE_ENTITIES];
 
@@ -597,7 +598,7 @@ void CL_ClearState (void)
 
 // wipe the entire cl structure
 	memset (&cl, 0, sizeof(cl));
-	memset (&cl_entities, 0, sizeof(cl_entities));
+	memset (cl_entities, 0, MAX_EDICTS * sizeof(cl_entities[0]));
 
 	SZ_Clear (&cls.netchan.message);
 
@@ -1405,6 +1406,7 @@ CL_InitLocal
 */
 void CL_InitLocal (void)
 {
+	cl_entities = Z_TagMallocFar (MAX_EDICTS * sizeof(cl_entities[0]), 0);
 	cls.state = ca_disconnected;
 	cls.realtime = Sys_Milliseconds ();
 
diff --git a/client/client.h b/client/client.h
index 2dc57a2..699d001 100644
--- a/client/client.h
+++ b/client/client.h
@@ -302,7 +302,7 @@ typedef struct
 	float	minlight;			// don't add when contributing less
 } cdlight_t;
 
-extern	centity_t	cl_entities[MAX_EDICTS];
+extern	centity_t	*cl_entities;
 extern	cdlight_t	cl_dlights[MAX_DLIGHTS];
 
 // the cl_parse_entities must be large enough to hold UPDATE_BACKUP frames of
diff --git a/client/ref.h b/client/ref.h
index 18ad121..31e56ee 100644
--- a/client/ref.h
+++ b/client/ref.h
@@ -22,7 +22,7 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 
 #define	MAX_DLIGHTS		32
 #define	MAX_ENTITIES	128
-#define	MAX_PARTICLES	4096
+#define	MAX_PARTICLES	1024
 #define	MAX_LIGHTSTYLES	256
 
 #define POWERSUIT_SCALE		4.0F
diff --git a/game/g_save.c b/game/g_save.c
index 1153f51..bf7c7aa 100644
--- a/game/g_save.c
+++ b/game/g_save.c
@@ -211,7 +211,7 @@ void InitGame (void)
 
 	// initialize all entities for this game
 	game.maxentities = maxentities->value;
-	g_edicts =  gi.TagMalloc (game.maxentities * sizeof(g_edicts[0]), TAG_GAME);
+	g_edicts =  gi.TagMallocFar (game.maxentities * sizeof(g_edicts[0]), TAG_GAME);
 	globals.edicts = g_edicts;
 	globals.max_edicts = game.maxentities;
 
@@ -503,7 +503,7 @@ void ReadGame (char *filename)
 		gi.error ("Savegame from an older version.\n");
 	}
 
-	g_edicts =  gi.TagMalloc (game.maxentities * sizeof(g_edicts[0]), TAG_GAME);
+	g_edicts =  gi.TagMallocFar (game.maxentities * sizeof(g_edicts[0]), TAG_GAME);
 	globals.edicts = g_edicts;
 
 	fread (&game, sizeof(game), 1, f);
diff --git a/game/game.h b/game/game.h
index 7b6218a..f8c4f04 100644
--- a/game/game.h
+++ b/game/game.h
@@ -174,6 +174,9 @@ typedef struct
 
 	// managed memory allocation
 	void	*(*TagMalloc) (int size, int tag);
+	// memory that is used less often and only written a word or halfword at
+	// a time, the Memory Expansion Pak of a DS ignores byte writes
+	void	*(*TagMallocFar) (int size, int tag);
 	void	(*TagFree) (void *block);
 	void	(*FreeTags) (int tag);
 
diff --git a/qcommon/common.c b/qcommon/common.c
index e147647..85773cf 100644
--- a/qcommon/common.c
+++ b/qcommon/common.c
@@ -19,6 +19,7 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
 // common.c -- misc functions used in client and server
 #include "qcommon.h"
+#include <malloc.h>
 #include <setjmp.h>
 
 #define	MAXPRINTMSG	4096
@@ -1160,12 +1161,12 @@ void Z_FreeTags (int tag)
 Z_TagMalloc
 ========================
 */
-void *Z_TagMalloc (int size, int tag)
+static void *Z_Alloc (int size, int tag, qboolean far)
 {
 	zhead_t	*z;
 	
 	size = size + sizeof(zhead_t);
-	z = malloc(size);
+	z = far ? far_malloc(size) : malloc(size);
 	if (!z)
 		Com_Error (ERR_FATAL, "Z_Malloc: failed on allocation of %i bytes",size);
 	memset (z, 0, size);
@@ -1183,6 +1184,20 @@ void *Z_TagMalloc (int size, int tag)
 	return (void *)(z+1);
 }
 
+void *Z_TagMalloc (int size, int tag)
+{
+	return Z_Alloc (size, tag, false);
+}
+
+// Memory that is used less often, like the game's edicts, goes to the far
+// heap (the Memory Expansion Pak of a DS) when there is one. That RAM
+// ignores byte writes, what is there is written a halfword or word at a
+// time, memcpy and memset handle the bytes at their ends.
+void *Z_TagMallocFar (int size, int tag)
+{
+	return Z_Alloc (size, tag, true);
+}
+
 /*
 ========================
 Z_Malloc
diff --git a/qcommon/files.c b/qcommon/files.c
index 4e2de9a..a90bd6c 100644
--- a/qcommon/files.c
+++ b/qcommon/files.c
@@ -453,7 +453,7 @@ pack_t *FS_LoadPackFile (char *packfile)
 	int				numpackfiles;
 	pack_t			*pack;
 	FILE			*packhandle;
-	dpackfile_t		info[MAX_FILES_IN_PACK];
+	dpackfile_t		*info;		// 256 KiB, too big for the stack
 	unsigned		checksum;
 
 	packhandle = fopen(packfile, "rb");
@@ -473,6 +473,7 @@ pack_t *FS_LoadPackFile (char *packfile)
 
 	newfiles = Z_Malloc (numpackfiles * sizeof(packfile_t));
 
+	info = Z_TagMallocFar (header.dirlen, 0);
 	fseek (packhandle, header.dirofs, SEEK_SET);
 	fread (info, 1, header.dirlen, packhandle);
 
@@ -490,6 +491,7 @@ pack_t *FS_LoadPackFile (char *packfile)
 		newfiles[i].filepos = LittleLong(info[i].filepos);
 		newfiles[i].filelen = LittleLong(info[i].filelen);
 	}
+	Z_Free (info);
 
 	pack = Z_Malloc (sizeof (pack_t));
 	strcpy (pack->filename, packfile);
diff --git a/qcommon/qcommon.h b/qcommon/qcommon.h
index 9481b4c..88c88f9 100644
--- a/qcommon/qcommon.h
+++ b/qcommon/qcommon.h
@@ -766,6 +766,7 @@ extern	int		time_after_ref;
 void Z_Free (void *ptr);
 void *Z_Malloc (int size);			// returns 0 filled memory
 void *Z_TagMalloc (int size, int tag);
+void *Z_TagMallocFar (int size, int tag);
 void Z_FreeTags (int tag);
 
 void Qcommon_Init (int argc, char **argv);
diff --git a/server/sv_game.c b/server/sv_game.c
index 2010122..07edd0a 100644
--- a/server/sv_game.c
+++ b/server/sv_game.c
@@ -367,6 +367,7 @@ void SV_InitGameProgs (void)
 	import.WriteAngle = PF_WriteAngle;
 
 	import.TagMalloc = Z_TagMalloc;
+	import.TagMallocFar = Z_TagMallocFar;
 	import.TagFree = Z_Free;
 	import.FreeTags = Z_FreeTags;
 
